src/service/data_products/filtering.py

```diff
@@ -3,26 +3,69 @@
 data products like genome attributes or samples.
 """
 
+from abc import ABC, abstractmethod
 from dateutil import parser
 from pydantic import BaseModel, Field
-from src.common.product_models.columnar_attribs_common_models import ColumnType
+from src.common.product_models.columnar_attribs_common_models import ColumnType, FilterStrategy
 
 from types import NotImplementedType
-from typing import Any, Self
+from typing import Annotated, Any, Self
 
 
 class SearchQueryPart(BaseModel):
+    variable_assignments: Annotated[dict[str, str] | None, Field(
+        description="A mapping of variable name to AQL expression. The variables must be assigned "
+            + "prior to the SEARCH operation. The variable names are expected to be unique "
+            + "across all filters.")
+    ] = None
     aql_lines: list[str] = Field(  # TDOO better docs once I figure out integration
         description="One or more lines of ArangoSearch AQL representing the filter")
     bind_vars: dict[str, Any] = Field(
-        description="The bind variables for the AQL lines")
+        description="The bind variables for the AQL lines. They are expected to be unique "
+            + "across all filters")
+
+
+class AbstractFilter(ABC):
+    """
+    The abstract base class for all filters.
+    """
+    
+    @abstractmethod
+    def from_string(self, type_: ColumnType, strategy: FilterStrategy, string: str) -> Self:
+        """
+        Parse the filter from a filter string. The syntax of the filter string is dependent
+        on the filter implementation.
+        
+        type_ - the column type to which the filter will apply.
+        strategy - the strategy for the filter.
+        string - the string to parse to create the filter.
+        """
+        raise NotImplementedError()
```

Comment:
Ah, not sure about this line. Why we need to raise NotImplementedError()? 

I thought a TypeError will be automatically raised if any of abstract methods is not implemented.

Reply:
Just to make it clear to the subclass implementer that there's no need to call the super method